internal/implementations/gateway/gateway.go

```diff
@@ -2,41 +2,65 @@ package implementation
 
 import (
 	"github.com/go-logr/logr"
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/events"
+	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/gateway-api/apis/v1alpha2"
 
 	"github.com/nginxinc/nginx-kubernetes-gateway/internal/config"
 	"github.com/nginxinc/nginx-kubernetes-gateway/pkg/sdk"
 )
 
 type gatewayImplementation struct {
-	conf config.Config
+	conf    config.Config
+	eventCh chan<- interface{}
 }
 
-func NewGatewayImplementation(conf config.Config) sdk.GatewayImpl {
+func NewGatewayImplementation(conf config.Config, eventCh chan<- interface{}) sdk.GatewayImpl {
 	return &gatewayImplementation{
-		conf: conf,
+		conf:    conf,
+		eventCh: eventCh,
 	}
 }
 
 func (impl *gatewayImplementation) Logger() logr.Logger {
 	return impl.conf.Logger
 }
 
-func (impl *gatewayImplementation) ControllerName() string {
-	return impl.conf.GatewayCtlrName
-}
-
 func (impl *gatewayImplementation) Upsert(gw *v1alpha2.Gateway) {
-	if gw.Name == impl.ControllerName() {
-		impl.Logger().Info("Found correct Gateway resource",
+	if gw.Namespace != impl.conf.GatewayNsName.Namespace || gw.Name != impl.conf.GatewayNsName.Name {
+		impl.Logger().Info("Gateway was upserted but ignored based on its namespace/name",
+			"namespace", gw.Namespace,
 			"name", gw.Name,
 		)
 		return
 	}
+
+	impl.Logger().Info("Gateway was upserted",
+		"namespace", gw.Namespace,
+		"name", gw.Name,
+	)
+
+	impl.eventCh <- &events.UpsertEvent{
+		Resource: gw,
+	}
 }
 
-func (impl *gatewayImplementation) Remove(key string) {
-	impl.Logger().Info("Gateway resource was removed",
-		"name", key,
+func (impl *gatewayImplementation) Remove(nsname types.NamespacedName) {
+	if nsname != impl.conf.GatewayNsName {
+		impl.Logger().Info("Gateway was removed but ignored based on its namespace/name",
+			"namespace", nsname.Namespace,
+			"name", nsname.Name,
+		)
+		return
+	}
+
+	impl.Logger().Info("Gateway was removed",
```

Comment:
Same issue with this log. We've begun the process of removing the gateway but we haven't finished. Is it guaranteed that the removal will succeed?

Reply:
relied to the previous similar comment